File: cpp/advanced-camera-features/connect-chunkdata/src/connect-chunkdata.cpp

#include <iostream>
#include <cstdlib>
#include <string>
#include <cstdio>
#include <vector>
#include <chrono>
#include <thread>

#include <ic4/ic4.h>

#include <console-helper.h>


struct PrintChunkExposureTimeListener : ic4::QueueSinkListener
{
	ic4::PropertyMap map_;
	ic4::PropFloat chunkExposureTime_;

	PrintChunkExposureTimeListener(ic4::PropertyMap m)
		: map_(m)
	{
		ic4::Error err;

		chunkExposureTime_ = map_.find(ic4::PropId::ChunkExposureTime, err);
		if (err.isError())
		{
			throw std::runtime_error("Float property ChunkExposureTime is not supported: " + err.message());
		}
	}

	void framesQueued(ic4::QueueSink& sink) override
	{
		// Do not throw from callback function, capture and log errors instead
		ic4::Error err;

		auto buffer = sink.popOutputBuffer(err);
		if (err.isError())
		{
			std::cerr << "popOutputBuffer failed: " << err.message() << std::endl;
			return;
		}

		// Use the image buffer as backend for read operations on chunk properties
		if (!map_.connectChunkData(buffer, err))
		{
			std::cerr << "connectChunkData failed: " << err.message() << std::endl;
			return;
		}
		
		// Read chunk property from image buffer
		auto val = chunkExposureTime_.getValue(err);
		if( err.isError() )
		{
			std::cerr << "ChunkExposureTime getValue failed: " << err.message() << std::endl;
			return;
		}

		// Disconnecting is not strictly necessary, but will release the buffer for reuse earlier
		map_.connectChunkData(nullptr, ic4::Error::Ignore());

		std::cout << " > ChunkExposureTime = " << val << std::endl;
	}
};

int main()
{
	// Initialize the library with sensible defaults:
	// - Throw exceptions on errors
	// - Log errors and warnings from API calls
	// - Log to stdout and Windows debug log
	ic4::InitLibraryConfig libraryConfig =
	{
		ic4::ErrorHandlerBehavior::Throw,
		ic4::LogLevel::Warning,
		ic4::LogLevel::Off,
		ic4::LogTarget::StdOut | ic4::LogTarget::WinDebug
	};
	ic4::initLibrary(libraryConfig);
	// Automatically call exitLibrary when returning from main
	std::atexit(ic4::exitLibrary);

	try
	{
		auto device_list = ic4::DeviceEnum::enumDevices();
		auto it = ic4_examples::console::select_from_list(device_list);
		if (it == device_list.end())
		{
			return -1;
		}

		ic4::Grabber grabber;
		grabber.deviceOpen(*it);

		auto map = grabber.devicePropertyMap();

		PrintChunkExposureTimeListener listener(map);
		auto sink = ic4::QueueSink::create(listener);

		// Try loading default UserSet to reset device to defaults
		// This reverses settings like TriggerMode=On, which would prevent this demo from running as expected
		map.setValue(ic4::PropId::UserSetSelector, "Default", ic4::Error::Ignore());
		map.executeCommand(ic4::PropId::UserSetLoad, ic4::Error::Ignore());

		// Chunkdata-related properties have to be configured before streamSetup,
		// because enabling them increases the payload size
		map.setValue(ic4::PropId::ChunkModeActive, true);
		map.setValue(ic4::PropId::ChunkSelector, "ExposureTime");
		// Ignore possible error, since the device might have ChunkEnable[ExposureTime] locked to true
		map.setValue(ic4::PropId::ChunkEnable, true, ic4::Error::Ignore());

		std::cout << "Configure resolution 640x480" << std::endl;
		map.setValue(ic4::PropId::Width, 640, ic4::Error::Ignore());
		map.setValue(ic4::PropId::Height, 480, ic4::Error::Ignore());

		std::cout << "Set AcquisitionFrameRate to 5" << std::endl;
		map.setValue(ic4::PropId::AcquisitionFrameRate, 5, ic4::Error::Ignore());

		std::cout << "Set ExposureAuto to Off" << std::endl;
		map.setValue(ic4::PropId::ExposureAuto, "Off");

		std::cout << "Set ExposureTime to 2 ms" << std::endl;
		map.setValue(ic4::PropId::ExposureTime, 2000);

		grabber.streamSetup(sink);

		std::cout << "Stream for 3 seconds" << std::endl;
		std::this_thread::sleep_for(std::chrono::seconds(3));

		std::cout << "Set ExposureTime to 8 ms" << std::endl;
		map.setValue(ic4::PropId::ExposureTime, 8000);

		std::cout << "Continue streaming for 3 seconds" << std::endl;
		std::this_thread::sleep_for(std::chrono::seconds(3));

		std::cout << "Set ExposureTime to 32 ms" << std::endl;
		map.setValue(ic4::PropId::ExposureTime, 32000);

		std::cout << "Continue streaming for 3 seconds" << std::endl;
		std::this_thread::sleep_for(std::chrono::seconds(3));

		// Stopping the stream is mandatory because we have the listener defined as a stack variable.
		// Just having everything go out of scope, the listener would be destroyed first and lead to undefined behavior.
		grabber.streamStop();

		return 0;
	}
	catch (const std::exception& ex)
	{
		std::cerr << "An exception occurred:" << std::endl;
		std::cerr << ex.what() << std::endl;
		return -10;
	}
}

File: python/advanced-camera-features/connect-chunkdata/connect-chunkdata.py
import time
import ic4


def example_connect_chunkdata():
    device_list = ic4.DeviceEnum.devices()
    for i, dev in enumerate(device_list):
        print(f"[{i}] {dev.model_name} ({dev.serial}) [{dev.interface.display_name}]")
    print(f"Select device [0..{len(device_list) - 1}]: ", end="")
    selected_index = int(input())
    dev_info = device_list[selected_index]

    grabber = ic4.Grabber()
    grabber.device_open(dev_info)

    prop_map = grabber.device_property_map

    # Create a QueueSink to get access to each received image buffer
    # The image buffers are processed in an instance of PrintChunkExposureTimeListener, defined below
    sink = ic4.QueueSink(PrintChunkExposureTimeListener(prop_map))

    # Make sure the selected device actually supports ChunkExposureTime
    try:
        _ = prop_map.find_float(ic4.PropId.CHUNK_EXPOSURE_TIME)
    except:
        print("ChunkExposureTime is not supported by this device")
        return

    # Reset all camera settings to default
    # This reverses settings like TriggerMode=On, which would prevent this demo from running as expected
    prop_map.set_value(ic4.PropId.USER_SET_SELECTOR, "Default")
    prop_map.execute_command(ic4.PropId.USER_SET_LOAD)

    # Chunkdata-related properties have to be configured before streamSetup,
    # because enabling them increases the payload size
    prop_map.set_value(ic4.PropId.CHUNK_MODE_ACTIVE, True)
    prop_map.set_value(ic4.PropId.CHUNK_SELECTOR, "ExposureTime")
    try:
        prop_map.set_value(ic4.PropId.CHUNK_ENABLE, True)
    except:
        # Ignore possible error, since the device might have ChunkEnable[ExposureTime] locked to true
        pass

    print("Configure resolution 640x480")
    prop_map.set_value(ic4.PropId.WIDTH, 640)
    prop_map.set_value(ic4.PropId.HEIGHT, 480)

    print("Set AcquisitionFrameRate to 5")
    prop_map.set_value(ic4.PropId.ACQUISITION_FRAME_RATE, 5)

    print("Set ExposureAuto to Off")
    prop_map.set_value(ic4.PropId.EXPOSURE_AUTO, "Off")

    print("Set ExposureTime to 2 ms")
    prop_map.set_value(ic4.PropId.EXPOSURE_TIME, 2000)

    grabber.stream_setup(sink)

    print("Stream for 3 seconds")
    time.sleep(3)

    print("Set ExposureTime to 8 ms")
    prop_map.set_value(ic4.PropId.EXPOSURE_TIME, 8000)

    print("Continue streaming for 3 seconds")
    time.sleep(3)

    print("Set ExposureTime to 32 ms")
    prop_map.set_value(ic4.PropId.EXPOSURE_TIME, 32000)

    print("Continue streaming for 3 seconds")
    time.sleep(3)


class PrintChunkExposureTimeListener(ic4.QueueSinkListener):
    prop_map: ic4.PropertyMap

    def __init__(self, prop_map: ic4.PropertyMap):
        self.prop_map = prop_map

    def sink_connected(self, sink: ic4.QueueSink, image_type: ic4.ImageType, min_buffers_required: int) -> bool:
        return True

    def frames_queued(self, sink: ic4.QueueSink):
        # Do not throw from callback function, capture and log errors instead
        try:
            buffer = sink.pop_output_buffer()

            # Use the image buffer as backend for read operations on chunk properties
            self.prop_map.connect_chunkdata(buffer)

            # Read chunk property from image buffer
            val = self.prop_map.get_value_float(ic4.PropId.CHUNK_EXPOSURE_TIME)
            print(f"ChunkExposureTime = {val}")

        except ic4.IC4Exception as ex:
            print(f"Error trying to request ChunkExposuretime: {ex.code} ({ex.message})")
        finally:
            # Disconnecting is not strictly necessary, but will release the buffer for reuse earlier
            self.prop_map.connect_chunkdata(None)

    def sink_disconnected(self, sink: ic4.QueueSink):
        pass


if __name__ == "__main__":
    ic4.Library.init(api_log_level=ic4.LogLevel.INFO, log_targets=ic4.LogTarget.STDERR)

    try:
        example_connect_chunkdata()
    finally:
        ic4.Library.exit()
